### DISCUSSION TITLE
Set custom RGB effects

### DISCUSSION THREAD

**Discussion by frannievas** [HomeSpan/HomeSpan#227] [Q&A]
Hi! Hope that everyone is having a great new year so far

I'm trying to develop custom effects for a ws2812b RGB led strip. My idea is to have 2-3 different effects and control them in the HomeKit app.

I've managed to program different effects following the [Example 12](https://github.com/HomeSpan/HomeSpan/tree/master/examples/10-RGB_LED).

I'm using the code from the example to get the RGB values and then use a library like FastLed to do the effects. Can I set up many effects like this?

**Comment by HomeSpan** [comment]
I have not used the FastLed library -- depending on how it is structured, it may be able to coexist with HomeSpan.  You'll need to test it to be sure.  By coincidence I am finalizing a new class in HomeSpan called "Pixel" that is designed to drive single-wire LED pixels and strips.  The prototype I used is the SK68 RGB LED since Espressif includes them on their ESP32-S2 and ESP32-C3 development boards.   The WS2812b uses the same protocol with some minor tweaks to configurable parameters related to the timings of High/Low pulses.   At present the class is designed for just a single LED, but I will be extending to drive multiple LEDs shortly.   So if the FastLed library gives you any problems, you may be able to use the Pixel class once complete (I have an initial version of the class deployed on the HomeSpan dev branch, but have not yet added any documentation).

The bigger question is how to integrate "effects" with the Home App.  At present, the Home App only allows control of color and brightness for a single LED.  You'll need to "repurpose"  another Home App Characteristic so that you can turn on/off special effects.  For example, you can repurpose the Brightness Characteristic, restrict its range to be between 1-5, and then have different value settings trigger different effects.   Unfortunately, there is no way of extending the Home App on the iPhone to provide customized controls.

**Reply by frannievas** [reply]
Regarding the question of how to integrate the effects. I saw this [video](https://youtu.be/2vgsBaVcC_Q) where they use a pair of VOCOlinc lights (which have custom RGB effects built by the manufacturer) and they are able to use the [EVE](https://apps.apple.com/us/app/eve-for-homekit/id917695792) app to setup custom scenes to use in the Home app. 

Maybe there's a way to implement the effects in the same way with HomeSpan, so the Eve app can see them and program this custom scenes

**Comment by HomeSpan** [comment]
Yes, that will works as well.  HomeSpan allows for custom characteristics (see the API for details).  I've used some with Eve for a similar purpose and its a good solution.  Note that Eve will display a generic interface for any custom characteristic it does not recognize (unlike the Home App, which just ignores them), provided the characteristic has its *description( field set to a descriptive name.  This field can be added to any characteristic using the `setDescription()` method.  This method is presently in the dev branch and will be released to the production branch in a few weeks.

**Comment by frannievas** [comment]
Amazing! I will start to look around the dev branch to see that new characteristic 👍 

And did you use Eve with this new feature or how? I didn't get it

**Comment by HomeSpan** [comment]
I've used custom Characteristics with Eve for other purposes, but not specifically for creating a mechanism to choose LED effects.  However, since Eve presents a generic slider, it can be used for just about anything.

**Comment by SnowballsChance** [comment]
I was inspired by this discussion and began experimenting. I have cobbled together an example of using HomeSpan's dimmable LED as the selector for multiple FastLED effects. 
I soon realised that different effects need differing update speeds so ended up running the FastLED animations  out of the main loop() and not in the dimmable LED's loop().

The code is rough but it works as an experiment. 

My experiment is available here: https://github.com/SnowballsChance/HS-Fastled-Brightness

**Comment by frannievas** [comment]
How nice that you are also interested in this topic @SnowballsChance. Is your repo private? I can't open it

**Reply by SnowballsChance** [reply]
@frannievas - Thanks for that - oops it was private. Now public.

**Comment by tprzyg** [comment]
I took a slightly different approach than @SnowballsChance, so the LED ribbon attached is controlled more or less like a single-color light strip (I struggled a bit with the HSV -> RGB conversion with FastLED using 0-255 vales rather than 0-360 and 0-100 respectively for H, S and V and also trueness of the converted color to what was being selected on the Home app. I also struggled with the dimming/brightness function, and in the end had to use full bigness in HSV->RGB conversion so that the dimming would work more or less as expected (rather than changing colors when the bigness was low). My code is here: https://github.com/tprzyg/FastLED-HomeSpan-WS2812B (repo should be public)

**Reply by counterbeing** [reply]
Very nice work! I'm going to try to adapt your code to another project. Thanks for sharing! 
How does it look in the apple home interface?

**Comment by HomeSpan** [comment]
HomeSpan version 1.5.0 has just been released and includes two new classes, **Pixel()** and **Dot()** that are optimized for controlling addressable one- and two-wire RGB and RGBW LEDs and LED strips.  See [HomeSpan Pixels](https://github.com/HomeSpan/HomeSpan/blob/master/docs/Pixels.md) for details.
